Error when passing empty list as argument to Lua function

First of all, thanks to the OpenTofu team for this cool and useful functionality!

I have had some success with using this provider, but there are some awkward failure cases as well. One of them was that it can handle an array being passed in, but not an empty array. This seems to be blocked by some validation in the Go code of the provider, rather than in the execution of my Lua code.

Also unrelated to this specific issue but of note in my snippet below: arrays (tables) seem to be 0-indexed not 1-indexed as they usually are in Lua?
I had to use `#arr == 0` and `return arr[0]` to get the correct behaviour - initially I tried `#arr == 1` and `return arr[1]` but that behaved incorrectly.

OpenTofu version: v1.8.1
terraform-provider-lua version: v0.0.2

Sample code:
```tf
locals {
  normalize_array = <<EOF
  -- If input array contains single value, return that value, otherwise return the original array.
  function normalize(arr)
    if type(arr) == "table" and #arr == 0 then
      return arr[0]
    end
    return arr
  end

  return normalize
  EOF
}

resource "vyos_config" "firewall_zones" {
  for_each = var.firewall.zones

  path = "firewall zone ${each.key}"
  value = jsonencode({
    foo1 = provider::lua::exec(local.normalize_array, ["a", "b"])
    foo2 = provider::lua::exec(local.normalize_array, ["c"])
    foo3 = provider::lua::exec(local.normalize_array, [])
  })
}
```

The error:

```
╷
│ Error: Error in function call
│ 
│   on modules/vyos/config/firewall.tf line 24, in resource "vyos_config" "firewall_zones":
│   24:     foo3 = provider::lua::exec(local.normalize_array, [])
│     ├────────────────
│     │ while calling provider::lua::exec(code, args...)
│     │ local.normalize_array is "function normalize(arr)\n  if type(arr) == \"table\" and #arr == 0 then\n    return arr[0]\n  end\n  return arr\nend\n\nreturn normalize"
│ 
│ Call to function "provider::lua::exec" failed: rpc error: code = Canceled desc = context canceled.
╵

Stack trace from the terraform-provider-lua_v0.0.2 plugin:

panic: unknown type implementation

goroutine 25 [running]:
github.com/zclconf/go-cty/cty.Type.MarshalJSON({{0x0?, 0x0?}})
        github.com/zclconf/go-cty@v1.13.1/cty/json.go:87 +0x647
github.com/zclconf/go-cty/cty/msgpack.(*dynamicVal).MarshalMsgpack(0xc0002d8200)
        github.com/zclconf/go-cty@v1.13.1/cty/msgpack/dynamic.go:18 +0x31
github.com/vmihailenco/msgpack/v5.marshalValue(0xc0002c05a0, {0xa51e60?, 0xc0002d8200?, 0x7f1475f60108?})
        github.com/vmihailenco/msgpack/v5@v5.4.1/encode_value.go:165 +0x11c
github.com/vmihailenco/msgpack/v5.(*Encoder).EncodeValue(0xc0002c05a0, {0xa51e60?, 0xc0002d8200?, 0x38?})
        github.com/vmihailenco/msgpack/v5@v5.4.1/encode.go:240 +0x7a
github.com/vmihailenco/msgpack/v5.(*Encoder).Encode(0xa323e0?, {0xa51e60?, 0xc0002d8200?})
        github.com/vmihailenco/msgpack/v5@v5.4.1/encode.go:226 +0x229
github.com/zclconf/go-cty/cty/msgpack.marshalDynamic(...)
        github.com/zclconf/go-cty@v1.13.1/cty/msgpack/marshal.go:208
github.com/zclconf/go-cty/cty/msgpack.marshal({{{0x0?, 0x0?}}, {0x0?, 0x0?}}, {{0xc1fd50?, 0x1103800?}}, {0x0, 0x0, 0x0}, 0xc0002c05a0)
        github.com/zclconf/go-cty@v1.13.1/cty/msgpack/marshal.go:53 +0x2d3
github.com/zclconf/go-cty/cty/msgpack.Marshal({{{0x0?, 0x0?}}, {0x0?, 0x0?}}, {{0xc1fd50?, 0x1103800?}})
        github.com/zclconf/go-cty@v1.13.1/cty/msgpack/marshal.go:37 +0x32b
main.CtyToProto({{{0x0?, 0x0?}}, {0x0?, 0x0?}})
        github.com/opentofu/terraform-provider-lua/main.go:329 +0x39
main.LuaToProto(0xc0002dc000?)
        github.com/opentofu/terraform-provider-lua/main.go:313 +0x2b
main.main.func1.2({0xc0002cf710, 0x2, 0x2})
        github.com/opentofu/terraform-provider-lua/main.go:283 +0x418
main.(*FunctionProvider).CallFunction(0xc00007d320, {0xc0002bd270?, 0xa68900?}, 0xc0002d4b70)
        github.com/opentofu/terraform-provider-lua/main.go:96 +0x17a
github.com/hashicorp/terraform-plugin-go/tfprotov6/tf6server.(*server).CallFunction(0xc0001a4140, {0xc1fb58?, 0xc0002d4570?}, 0xc0002bd220)
        github.com/hashicorp/terraform-plugin-go@v0.22.1/tfprotov6/tf6server/server.go:987 +0x623
github.com/hashicorp/terraform-plugin-go/tfprotov6/internal/tfplugin6._Provider_CallFunction_Handler({0xb001c0, 0xc0001a4140}, {0xc1fb58, 0xc0002d4570}, 0xc0002ab000, 0x0)
        github.com/hashicorp/terraform-plugin-go@v0.22.1/tfprotov6/internal/tfplugin6/tfplugin6_grpc.pb.go:608 +0x1a6
google.golang.org/grpc.(*Server).processUnaryRPC(0xc000103000, {0xc1fb58, 0xc0002d44e0}, {0xc23ec0, 0xc0000d9860}, 0xc0002c9d40, 0xc0002412c0, 0x1095090, 0x0)
        google.golang.org/grpc@v1.62.1/server.go:1386 +0xdf8
google.golang.org/grpc.(*Server).handleStream(0xc000103000, {0xc23ec0, 0xc0000d9860}, 0xc0002c9d40)
        google.golang.org/grpc@v1.62.1/server.go:1797 +0xe87
google.golang.org/grpc.(*Server).serveStreams.func2.1()
        google.golang.org/grpc@v1.62.1/server.go:1027 +0x8b
created by google.golang.org/grpc.(*Server).serveStreams.func2 in goroutine 16
        google.golang.org/grpc@v1.62.1/server.go:1038 +0x125

Error: The terraform-provider-lua_v0.0.2 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.
```


P.S. Would love to see a Python provider in future as well one day, as I'm much more familiar with it than Lua or Go 😄

--- a/lib.lua
+++ b/lib.lua
@@ -5,3 +5,14 @@ end
 function echo(value)
     return value
 end
+
+-- If input array contains single value, return that value, otherwise return the original array.
+function normalize(arr)
+  if type(arr) == "table" and next(arr) == nil then
+    return {}
+  end
+  if type(arr) == "table" and #arr == 0 then
+    return arr[0]
+  end
+  return arr
+end

--- a/main.go
+++ b/main.go
@@ -392,12 +392,12 @@ func CtyToLua(arg cty.Value, l *lua.State) error {
 
 func LuaToCty(l *lua.State) (cty.Value, error) {
 	if l.IsNone(-1) {
-		return cty.NilVal, fmt.Errorf("none value should not be returned")
+		return cty.NullVal(cty.DynamicPseudoType), fmt.Errorf("none value should not be returned")
 	}
 
 	switch t := l.TypeOf(-1); t {
 	case lua.TypeNil:
-		return cty.NilVal, nil
+		return cty.NullVal(cty.DynamicPseudoType), nil
 	case lua.TypeBoolean:
 		return cty.BoolVal(l.ToBoolean(-1)), nil
 	case lua.TypeNumber:
@@ -421,7 +421,7 @@ func LuaToCty(l *lua.State) (cty.Value, error) {
 			// Decode key (also modifies)
 			key, ok := l.ToString(-1)
 			if !ok {
-				return cty.NilVal, fmt.Errorf("bad table index")
+				return cty.NullVal(cty.DynamicPseudoType), fmt.Errorf("bad table index")
 			}
 
 			l.Pop(1)
@@ -453,8 +453,8 @@ func LuaToCty(l *lua.State) (cty.Value, error) {
 				return cty.ObjectVal(mv), nil
 			}
 		}
-		return cty.ListVal(av), nil
+		return cty.TupleVal(av), nil
 	default:
-		return cty.NilVal, fmt.Errorf("unhanded return type %s!", t)
+		return cty.NullVal(cty.DynamicPseudoType), fmt.Errorf("unhanded return type %s!", t)
 	}
 }

--- a/main.tf
+++ b/main.tf
@@ -18,3 +18,15 @@ provider "tester" {
 output "test" {
 	value = provider::tester::echo(tomap({"foo": {"bar": 190}}))
 }
+
+output "norm_0" {
+  value = provider::tester::normalize([])
+}
+
+output "norm_1" {
+  value = provider::tester::normalize([1])
+}
+
+output "norm_2" {
+  value = provider::tester::normalize([1, {"foo": "bar"}])
+}

